core/src/Streamly/Internal/Data/Stream/TH.hs

```diff
@@ -0,0 +1,542 @@
+-- |
+-- Module      : Streamly.Internal.Data.Stream.TH
+-- Copyright   : (c) 2022 Composewell Technologies
+-- License     : BSD-3-Clause
+-- Maintainer  : streamly@composewell.com
+-- Stability   : experimental
+-- Portability : GHC
+--
+-- Template haskell helpers for dynamically creating types.
+--
+-- >>> import Language.Haskell.TH
+-- >>> import Streamly.Internal.Data.Stream.TH
+-- >>> expr <- runQ (makeStreamMonad "Parallel" "bindConcurrent" True)
+-- >>> putStrLn $ pprint expr
+-- newtype Parallel m a = Parallel (Stream m a)
+-- mkParallel :: Stream m a -> Parallel m a
+-- mkParallel = Parallel
+-- unParallel :: Parallel m a -> Stream m a
+-- unParallel (Parallel strm) = strm
+-- instance Monad m => Functor (Parallel m)
+--     where {-# INLINE fmap #-}
+--           fmap f (Parallel strm) = Parallel (fmap f strm)
+-- instance Streamly.Internal.Data.Stream.Concurrent.MonadAsync m => Monad (Parallel m)
+--     where {-# INLINE (>>=) #-}
+--           (>>=) (Parallel strm1) f = let f1 a = unParallel (f a)
+--                                       in Parallel (bindConcurrent strm1 f1)
+-- instance Streamly.Internal.Data.Stream.Concurrent.MonadAsync m => Applicative (Parallel m)
+--     where {-# INLINE pure #-}
+--           pure = Parallel . Streamly.Data.Stream.fromPure
+--           {-# INLINE (<*>) #-}
+--           (<*>) (Parallel strm1) (Parallel strm2) = Parallel (Control.Monad.ap strm1 strm2)
+-- instance MonadTrans Parallel
+--     where {-# INLINE lift #-}
+--           lift = Parallel . lift
+-- instance (Monad (Parallel m), MonadIO m) => MonadIO (Parallel m)
+--     where {-# INLINE liftIO #-}
+--           liftIO = Parallel . (lift . liftIO)
+-- instance (Monad (Parallel m),
+--           MonadThrow m) => MonadThrow (Parallel m)
+--     where {-# INLINE throwM #-}
+--           throwM = Parallel . (lift . throwM)
+-- instance (Monad (Parallel m), MonadReader r m) => MonadReader r
+--                                                               (Parallel m)
+--     where {-# INLINE ask #-}
+--           ask = lift ask
+--           {-# INLINE local #-}
+--           local f (Parallel strm) = Parallel (local f strm)
```

Comment:
Can we push this example to mkStreamMonad function itself? And have a similar example for the Zip Applicative macro in the corresponding function.